cli/cmd/awsiam.go

```diff
@@ -0,0 +1,304 @@
+		return nil // we didn't create this role, we should not delete it
+	}
+
+	// List managed policies attached to this role (assume there are no inline policies)
+	listInput := &iam.ListAttachedRolePoliciesInput{
+		RoleName: role.RoleName,
+	}
+	listOutput, err := c.ListAttachedRolePolicies(context.Background(), listInput)
+	if err != nil {
+		return err
+	}
+
+	// Detach managed policies
+	for _, attachedPolicy := range listOutput.AttachedPolicies {
+		cli.Log.Debugw("detaching policy", "policy", attachedPolicy, "role", role)
+		detachInput := &iam.DetachRolePolicyInput{
+			PolicyArn: attachedPolicy.PolicyArn,
+			RoleName:  role.RoleName,
+		}
+		_, err := c.DetachRolePolicy(context.Background(), detachInput)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Delete the role
+	cli.Log.Debugw("deleting role", "role", role)
+	deleteRoleInput := &iam.DeleteRoleInput{
+		RoleName: role.RoleName,
+	}
+	_, err = c.DeleteRole(context.Background(), deleteRoleInput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setupSSMRole sets up an IAM role for SSM to assume.
+// If `roleName` is not the empty string, then use that role
+// instead of creating a new one.
+func setupSSMRole(cfg aws.Config, roleName string) (types.Role, error) {
+	if roleName != "" {
+		return getRoleFromName(cfg, roleName)
+	} else { // user did not provide a role, creating one now
+		return createSSMRole(cfg)
+	}
+}
+
+func getRoleFromName(cfg aws.Config, roleName string) (types.Role, error) {
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	input := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
+	}
+	output, err := c.GetRole(context.Background(), input)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	return *output.Role, nil
+}
+
+// createSSMRole makes a call to the AWS API to create an IAM role.
+// This role allows the current user to assume it.
+// Returns information about the newly created role and any errors.
+func createSSMRole(cfg aws.Config) (types.Role, error) {
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	// Check if role already exists (in case CLI is run after interrupt or error)
+	getInput := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
+	}
+	getOutput, err := c.GetRole(context.Background(), getInput)
+	if err == nil && getOutput.Role != nil {
+		return *getOutput.Role, err // we previously created the role, use it
+	}
+
+	trustPolicyDocument := `{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Effect": "Allow",
+			"Principal": { "Service": "ec2.amazonaws.com" },
+			"Action": "sts:AssumeRole"
+		}
+	]
+}`
+	input := &iam.CreateRoleInput{
+		AssumeRolePolicyDocument: &trustPolicyDocument,
+		RoleName:                 aws.String(roleName),
+		Description: aws.String(
+			`Ephemeral role to install Lacework agents using SSM; created by the Lacework CLI.
+Safe to delete if found`,
+		),
+	}
```

Comment:
It's polite to add tags that clearly identify this role as lacework-derived ephemeral for automation. There are tools that folks use that specifically look at tags, so this can surface things for folks that just having a good name doesn't (untested code):

```suggestion
	input := &iam.CreateRoleInput{
		AssumeRolePolicyDocument: &trustPolicyDocument,
		RoleName:                 aws.String(roleName),
		Description: aws.String(
			`Ephemeral role to install Lacework agents using SSM; created by the Lacework CLI.
Safe to delete if found`,
		),
		Tags: []*iam.Tag{
		{
		    Key:   aws.String("LaceworkAutomation"),
		    Value: aws.String("agent-ssm-install")
		}
		}
	}
```

Reply:
Good idea, done with bffd96d7613cbc53e83f996f0b922ae2b2f6c74b